source/helpers/components/anchors.ts

```diff
@@ -0,0 +1,38 @@
+import Token from 'markdown-it/lib/token';
+import type anchor from 'markdown-it-anchor';
+
+// Custom permalink function, inspired by and modification of linkInsideHeader
+// More context on custom permalink fns: https://github.com/valeriangalliat/markdown-it-anchor#custom-permalink
+// linkInsideHeader function: https://github.com/valeriangalliat/markdown-it-anchor/blob/649582d58185b00cfb2ceee9b6b4cd6aafc645b7/permalink.js#L76
+export const renderPermalink: anchor.PermalinkGenerator = (
+  slug,
+  opts: anchor.LinkInsideHeaderPermalinkOptions,
+  state,
+  idx,
+) => {
+  opts.ariaHidden = false;
+
+  const title = state?.tokens[idx + 1].children
+    ?.filter(
+      (token: Token) => token.type === 'text' || token.type === 'code_inline',
+    )
+    .reduce((acc, t) => acc + t.content, '');
+
+  const linkTokens = [
+    Object.assign(new state.Token('link_open', 'a', 1), {
+      attrs: [
+        ['class', 'anchor'],
+        ['href', `#${slug}`],
+        ...(opts.ariaHidden ? [['aria-hidden', 'true']] : []),
+      ],
+    }),
+    Object.assign(new state.Token('html_inline', '', 0), {
+      content: `<span class="visuallyhidden">${
+        title ? title : 'section'
+      } permalink'</span>`,
+    }),
+    new state.Token('link_close', 'a', -1),
+  ];
+
+  state?.tokens[idx + 1]?.children?.push(...linkTokens);
+};
```

Comment:
@sanajaved7 I haven't fully parsed this yet, but I'm curious why we need a more custom/complex solution vs the approach used in OddSite? https://github.com/oddbird/oddleventy/blob/main/src/filters/type.cjs#L19-L29

Reply:
@jgerigmeyer good question! I ultimately went this route to leverage as much of the existing styles and to match the existing implementation of anchor links as much as possible. I tried using both `linkAfterHeader` and `linkInsideHeader` but they both fell short and didn't quite do everything I wanted. With `linkAfterHeader`, I could customize the assistive text and access the post title but the anchor link symbol on hover did not work. And with `linkInsideHeader`,  I could customize the symbol content to provide the `span` with assistive text but I couldn't access the post title data since that's only available in the `assistiveText` method. You'll notice this custom method looks a lot like the original `linkInsideHeader` method but I'm adjusting it to actually [access](https://github.com/oddbird/sass-site/pull/37/files#diff-b8a7ef5e455f5f50a445f98f4daf8c75e8541b178465e26741922870bd762be5R15) the `title` value and [use it](https://github.com/oddbird/sass-site/pull/37/files#diff-b8a7ef5e455f5f50a445f98f4daf8c75e8541b178465e26741922870bd762be5R31).

Reply:
@sanajaved7 I see! This seems okay if the primary goal is retaining the current markup, but given the [warnings about a11y issues](https://github.com/valeriangalliat/markdown-it-anchor#link-inside-header) with `linkInsideHeader`, I wonder if we want to recommend some slight changes that could simplify this. @mirisuzanne or @stacyk, what do you think?

Reply:
I had some questions/doubts on the accessibility here, too. This [post](https://amberwilson.co.uk/blog/are-your-anchor-links-accessible/#a-word-of-caution) has a lot of context on anchor link accessibility and links to the discussion/issue that were opened in the markdown-it-anchor repo as well. 

I was curious if some of the the accessibility issues with this approach (link inside header) were addressed if the symbol to the anchor link only appears on hover and thus wouldn't be read out by screen readers. I'm not sure if that's enough since the issue with the "[title] permalink" being read as a heading — that the post linked above [mentions](https://amberwilson.co.uk/blog/are-your-anchor-links-accessible/#a-word-of-caution) here — still exists though. Definitely would appreciate some guidance from others on this.

Reply:
I just tested the site with voiceover and it read it ok while navigating with my keyboard. When I look at the inspector's accessibility panel in Chrome (and Firefox), I don't know if the text repeated in these headings will be repeated in any other assistive technology:

![Screenshot 2023-04-25 at 11 40 11 PM](https://user-images.githubusercontent.com/1581694/234471901-d258fb66-d110-4382-b098-7d5f89d08533.jpg)

I can dig into this more tomorrow